Update to support gcc-7-20161002

ibuclaw opened the pull request:
It's been a few months, and snapshot tarballs eventually get recycled.

ibuclaw committed:
Update to support gcc-7-20161002

ibuclaw commented on gcc/d/patches/patch-versym-os-7.x:
```diff
@@ -196,25 +196,6 @@ These official OS versions are not implemented:
  #define GNU_USER_DYNAMIC_LINKER32      GLIBC_DYNAMIC_LINKER32
  #define GNU_USER_DYNAMIC_LINKER64      GLIBC_DYNAMIC_LINKER64
---- a/gcc/config/knetbsd-gnu.h
-+++ b/gcc/config/knetbsd-gnu.h
```
knetbsd has been dropped from gcc.

jpf91 commented:
> Yeah, I was trying to diagnose the phantom failure in #236. I have no idea why it fails... yet. There must be some environment setting inside expect/dejagnu that isn't when compiling + running by hand.

That's really strange indeed. I can't reproduce the problem locally at all and on semaphore CI it's only reproducible when running through dejagnu, but not with manual compilation. And there's no useful output in the `gdc.log` file either...

ibuclaw commented:
Giving gcc-7-20161002 a try.

ibuclaw commented:
Latest snapshot has a new change that removed unreferenced dwarf DIEs. Subsequently was getting an ICE in gdc with code more or less like this.

```
int outer()
{
  struct type
  {
    int inner() { return 0; }
  }
  return new type().inner();
}
```

Although debug information was generated for both `outer` and `type`, and context hierarchy was correct, the `TYPE_DECL` for `type` was not referenced as being used by anywhere in the GCC trees.  Corrective action was to use `d_pushdecl` to put it in the `BLOCK_VARS` of `outer`.

Have changed some things around so that `d_add_global_declaration` has been collapsed into `d_pushdecl` (some decls were also being added to the global declarations list twice).  However have subsequently been getting other ICE's, which may or may not be related to the change. :-)

ibuclaw commented:
Note to self, more complex example.

```
struct ModuleGroup
{
    void sortCtors()
    {
        struct StackRec
        {
            immutable(int*)[] mods;
        }
        assert(0);
    }

}

struct DSO
{
    static DSO*[] loadedDSOs;

    static opApply(int delegate(DSO) dg)
    {
        foreach (dso; loadedDSOs)
            dg(*dso);
        return 0;
    }

    ModuleGroup moduleGroup;
}


void rt_moduleCtor()
{
    foreach (sg; DSO)
        sg.moduleGroup.sortCtors;
}
```

ibuclaw commented:
ICE happens because `__xopEquals` is marked as inline, but perhaps it shouldn't be.

ibuclaw commented:
Another, definitely caused by the change I've made in the frontend (`d_pushdecl`).

```
class LocalTime 
{
    this() immutable { }

    static localTime = new immutable(LocalTime);
}
```

Where we overwrite the `TYPE_DECL` with an `IDENTIFIER_NODE` in `d_finish_symbol`.

```
if (TREE_TYPE (decl) == unknown_type_node)
  {
    TREE_TYPE (decl) = TREE_TYPE (sinit);  // sinit.type == LocalTime.
    TYPE_NAME (TREE_TYPE (decl)) = get_identifier (sym->Sident)
  }
```

Good, I am exposing hidden bugs in our code generation. :-)

ibuclaw commented:
Variation on the first (there's an extra binding level).

```
int outer()
{
  { 
    struct type
    { 
      int inner() { return 0; }
    }
  }
  return 0;
}
```

ibuclaw commented:
Fixed by marking the bind_expr returned from `finish_scope` as having side effects (it contains BLOCK_VARS, so it certainly does).

ibuclaw commented:
The reason looks like a segfault for the failure in test10.d

```
Testing runnable/test10.d, -shared-libphobos 
doing compile
pid is 24419 -24419
output is  status 0
spawning command ./test10.exe 
exp9 file8
close result is child killed: segmentation violation
FAIL: runnable/test10.d -shared-libphobos  execution test
Testing runnable/test10.d, -g -shared-libphobos 
doing compile
pid is 24435 -24435
output is  status 0
spawning command ./test10.exe 
exp9 file8
close result is child killed: segmentation violation
FAIL: runnable/test10.d -g -shared-libphobos  execution test
Testing runnable/test10.d, -O2 -shared-libphobos 
doing compile
pid is 24451 -24451
output is  status 0
spawning command ./test10.exe 
exp9 file8
```

ibuclaw commented:
And I can avoid getting a segfault I set `LD_RUN_PATH`

```
export LD_RUN_PATH=/foo
```

That is interesting.

When running in verbose mode, the only difference that should make is this:

```
-LD_RUN_PATH=.:/home/runner/GDC/build/x86_64-pc-linux-gnu/./libphobos/src/.libs:/home/runner/GDC/build/x86_64-pc-linux-gnu/./libphobos/libdruntime/.libs:/home/runner/GDC/build/x86_64-pc-linux-gnu/./libstdc++-v3/s
rc/.libs:/home/runner/GDC/build/gcc:/foo
+LD_RUN_PATH=.:/home/runner/GDC/build/x86_64-pc-linux-gnu/./libphobos/src/.libs:/home/runner/GDC/build/x86_64-pc-linux-gnu/./libphobos/libdruntime/.libs:/home/runner/GDC/build/x86_64-pc-linux-gnu/./libstdc++-v3/s
rc/.libs:/home/runner/GDC/build/gcc
```

ibuclaw committed:
Ensure that the writable data segment is aligned.

Because ELF specifies that p_vaddr need not be aligned, just that (p_vaddr & p_align) == (p_offset & p_align).

ibuclaw committed:
Update to support gcc-7-20161030
